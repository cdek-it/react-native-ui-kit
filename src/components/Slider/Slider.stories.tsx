import type { Meta, StoryObj } from '@storybook/react'
import { useState } from 'react'
import { Text, View } from 'react-native'

import { Slider, type SliderProps } from './Slider'

const meta: Meta<typeof Slider> = {
  title: 'Slider',
  component: Slider,
  args: { disabled: false },
  render: (args) => Template(args),
}

const Template = (args: SliderProps) => {
  const [positionMin, setPositionMin] = useState<number>(40)

  const [positionMinRange, setPositionMinRange] = useState<number>(20)
  const [positionMaxRange, setPositionMaxRange] = useState<number>(80)

  const handleChangeMin = (value: number) => {
    setPositionMin(value)
  }

  const handleChangeMinRange = (value: number) => {
    setPositionMinRange(value)
  }

  const handleChangeMaxRange = (value: number) => {
    setPositionMaxRange(value)
  }

  return (
    <View style={{ gap: 20 }}>
      <View style={{ gap: 10 }}>
        <Text>{positionMinRange.toFixed(0)}</Text>
        <Text>{positionMaxRange.toFixed(0)}</Text>
        <Slider
          {...args}
          range
          maxPointerValueInit={80}
          minPointerValueInit={20}
          onMaxPointerValueChange={(value) => handleChangeMaxRange(value)}
          onMinPointerValueChange={(value) => handleChangeMinRange(value)}
        />
      </View>

      <View style={{ gap: 10 }}>
        <Text>{positionMin.toFixed(0)}</Text>
        <Slider
          {...args}
          minPointerValueInit={40}
          range={false}
          onMinPointerValueChange={(value) => handleChangeMin(value)}
        />
      </View>
    </View>
  )
}

export default meta

type Story = StoryObj<typeof Slider>

const SliderStory: Story = { args: {} }
export { SliderStory as Slider }
