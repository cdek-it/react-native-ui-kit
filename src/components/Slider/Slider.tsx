import { memo, useCallback, useMemo, useState } from 'react'
import {
  type AccessibilityProps,
  type LayoutChangeEvent,
  View,
  type ViewProps,
} from 'react-native'
import { Gesture, GestureDetector } from 'react-native-gesture-handler'
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  interpolate,
  Extrapolation,
  runOnJS,
} from 'react-native-reanimated'

import { makeStyles } from '../../utils/makeStyles'

export interface SliderProps
  extends AccessibilityProps,
    Pick<ViewProps, 'testID'> {
  /**
   * Управление доступностью компонента
   * @default false
   */
  disabled?: boolean
  /**
   * Признак наличия диапазона выбираемых значений
   * @default false
   */
  range?: boolean
  /**
   * Начальное значение/позиция стартового ползунка
   * @default 0
   */
  minPointerValueInit: number
  /**
   * Конечное значение/позиция конечного ползунка
   * @default 100
   */
  maxPointerValueInit?: number
  /**
   * Значение/позиция возвращаемое стартовым ползунком
   */
  onMinPointerValueChange: (value: number) => void
  /**
   * Значение/позиция возвращаемое конечным ползунком
   */
  onMaxPointerValueChange: (value: number) => void
}

const MIN_TRACK_SCALE = 0
const MAX_TRACK_SCALE = 100

const clamp = (val: number, min: number, max: number) => {
  'worklet'

  return Math.min(Math.max(val, min), max)
}

/**
 * Используется для указания значения или диапазона значений с помощью ползунка
 * @see https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-6090&m=dev
 */
export const Slider = memo<SliderProps>(
  ({
    disabled = false,
    range = false,
    minPointerValueInit = 0,
    maxPointerValueInit = 100,
    onMinPointerValueChange,
    onMaxPointerValueChange,
    testID,
    ...rest
  }) => {
    const styles = useStyles()
    const minPointX = useSharedValue(0)
    const maxPointX = useSharedValue(0)

    const trackWidth = useSharedValue(0)
    const pointerWidth = styles.point.width

    const [isPressed, setIsPressed] = useState(false)

    const prevMinPointX = useSharedValue(0)
    const prevMaxPointX = useSharedValue(0)

    const pointerStyle = useMemo(
      () => [
        styles.point,
        isPressed && styles.hovered,
        disabled && styles.disabledView,
      ],
      [styles.point, isPressed, styles.hovered, styles.disabledView, disabled]
    )

    const lineStyle = useMemo(
      () => [
        styles.line,
        isPressed && styles.hovered,
        disabled && styles.disabledView,
      ],
      [styles.line, isPressed, styles.hovered, styles.disabledView, disabled]
    )

    const interpolateInitVal = useCallback(
      (value: number, width: number) => {
        return interpolate(
          value,
          [MIN_TRACK_SCALE, MAX_TRACK_SCALE],
          [0, width - pointerWidth],
          Extrapolation.CLAMP
        )
      },
      [pointerWidth]
    )

    const onContainerLayout = useCallback(
      (event: LayoutChangeEvent) => {
        event.target.measure((x, y, width) => {
          trackWidth.value = width
          const min = interpolateInitVal(minPointerValueInit, width)
          const max = interpolateInitVal(maxPointerValueInit, width)

          minPointX.value = min
          maxPointX.value = max
        })
      },
      [
        minPointerValueInit,
        maxPointerValueInit,
        interpolateInitVal,
        trackWidth,
        minPointX,
        maxPointX,
      ]
    )

    const returnMinVal = useCallback(
      (value: number) => {
        const min = interpolate(
          value,
          [0, trackWidth.value - pointerWidth * 2],
          [MIN_TRACK_SCALE, MAX_TRACK_SCALE],
          Extrapolation.CLAMP
        )
        onMinPointerValueChange(min)
      },
      [trackWidth, pointerWidth, onMinPointerValueChange]
    )

    const returnMaxVal = useCallback(
      (value: number) => {
        const min = interpolate(
          value,
          [pointerWidth, trackWidth.value - pointerWidth],
          [MIN_TRACK_SCALE, MAX_TRACK_SCALE],
          Extrapolation.CLAMP
        )
        onMaxPointerValueChange(min)
      },
      [trackWidth, pointerWidth, onMaxPointerValueChange]
    )

    const panMinPoint = Gesture.Pan()
      .minDistance(1)
      .onBegin(() => {
        prevMinPointX.value = minPointX.value
        runOnJS(setIsPressed)(true)
      })
      .onUpdate((event) => {
        const maxTranslateX = trackWidth.value - pointerWidth

        const minPointPosition = clamp(
          prevMinPointX.value + event.translationX,
          0,
          range ? maxPointX.value - pointerWidth : maxTranslateX
        )

        minPointX.value = minPointPosition
        runOnJS(returnMinVal)(minPointPosition)
      })
      .onFinalize(() => {
        runOnJS(setIsPressed)(false)
      })

    const minPointStyle = useAnimatedStyle(() => ({
      transform: [{ translateX: minPointX.value }],
    }))

    const panMaxPoint = Gesture.Pan()
      .minDistance(1)
      .onBegin(() => {
        prevMaxPointX.value = maxPointX.value
        runOnJS(setIsPressed)(true)
      })
      .onUpdate((event) => {
        const maxTranslateX = trackWidth.value - pointerWidth

        const maxPointPosition = clamp(
          prevMaxPointX.value + event.translationX,
          minPointX.value + pointerWidth,
          maxTranslateX
        )
        maxPointX.value = maxPointPosition
        runOnJS(returnMaxVal)(maxPointPosition)
      })
      .onFinalize(() => {
        runOnJS(setIsPressed)(false)
      })

    const maxPointStyle = useAnimatedStyle(() => ({
      transform: [{ translateX: maxPointX.value }],
    }))

    const minLineStyle = useAnimatedStyle(() => ({
      left: 0,
      width: minPointX.value + pointerWidth / 2,
    }))

    const betweenLineStyle = useAnimatedStyle(() => ({
      left: minPointX.value + pointerWidth / 2,
      width: maxPointX.value - minPointX.value,
    }))

    return (
      <View
        style={styles.container}
        testID={testID}
        onLayout={onContainerLayout}
        {...rest}
      >
        <View style={styles.track}>
          {range ? null : ( // индикатор старта
            <Animated.View style={[lineStyle, minLineStyle]} />
          )}

          {range ? ( // индикатор между точками
            <Animated.View style={[lineStyle, betweenLineStyle]} />
          ) : null}

          <GestureDetector gesture={panMinPoint}>
            <Animated.View
              pointerEvents={disabled ? 'none' : 'auto'}
              style={[pointerStyle, minPointStyle]}
            />
          </GestureDetector>

          {range ? ( // индикатор между точками
            <GestureDetector gesture={panMaxPoint}>
              <Animated.View
                pointerEvents={disabled ? 'none' : 'auto'}
                style={[pointerStyle, maxPointStyle]}
              />
            </GestureDetector>
          ) : null}
        </View>
      </View>
    )
  }
)

const useStyles = makeStyles(({ theme }) => {
  return {
    container: {
      width: '100%',
      justifyContent: 'center',
      alignItems: 'center',
    },
    track: {
      height: theme.Form.Slider.sliderHorizontalHeight,
      backgroundColor: theme.Form.Slider.sliderBg,
      borderRadius: theme.Form.Slider.sliderHandleBorderRadius,
      position: 'relative',
      width: '100%',
      justifyContent: 'center',
    },

    line: {
      height: theme.Form.Slider.sliderHorizontalHeight,
      borderRadius: theme.Form.Slider.sliderHandleBorderRadius,
      position: 'absolute',
      backgroundColor: theme.Form.Slider.sliderRangeBg,
    },

    point: {
      width: theme.Form.Slider.sliderHandleWidth,
      height: theme.Form.Slider.sliderHandleHeight,
      borderRadius: theme.Form.Slider.sliderHandleHeight / 2,
      position: 'absolute',
      backgroundColor: theme.Form.Slider.sliderHandleBg,
    },

    disabledView: { backgroundColor: 'grey' },

    hovered: { backgroundColor: theme.Form.Slider.sliderHandleHoverBg },
  }
})
