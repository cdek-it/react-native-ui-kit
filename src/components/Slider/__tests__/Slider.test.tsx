import { render } from '@testing-library/react-native'

import { Slider, type SliderProps } from '../Slider'

describe('Slider', () => {
  const defaultProps: SliderProps = {
    minPointerValueInit: 10,
    onMaxPointerValueChange: (_) => undefined,
    onMinPointerValueChange: (_) => undefined,
  }

  describe('snapshots', () => {
    const snapshotCases: Array<[string, Partial<SliderProps>]> = [
      ['default', {}],
      ['disabled = false, range = false', { disabled: false, range: false }],
      ['disabled = true, range = false', { disabled: true, range: false }],

      ['disabled = false, range = true', { disabled: false, range: true }],

      ['disabled = true, range = true', { disabled: true, range: true }],

      [
        'disabled = false, range = true, maxPointerValueInit = 80',
        { disabled: false, range: true, maxPointerValueInit: 80 },
      ],

      [
        'disabled = true, range = true, maxPointerValueInit = 80',
        { disabled: true, range: true, maxPointerValueInit: 80 },
      ],
    ]

    test.each(snapshotCases)('%s', (_, props) => {
      const { toJSON } = render(<Slider {...defaultProps} {...props} />)

      expect(toJSON()).toMatchSnapshot()
    })
  })
})
