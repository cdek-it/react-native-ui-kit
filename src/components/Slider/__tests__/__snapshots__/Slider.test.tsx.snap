// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider snapshots default 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 14,
        "height": 3.5,
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "left": 0,
            "width": 8.75,
          },
        }
      }
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 14,
          "height": 3.5,
          "left": 0,
          "position": "absolute",
          "width": 8.75,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
  </View>
</View>
`;

exports[`Slider snapshots disabled = false, range = false 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 14,
        "height": 3.5,
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "left": 0,
            "width": 8.75,
          },
        }
      }
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 14,
          "height": 3.5,
          "left": 0,
          "position": "absolute",
          "width": 8.75,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
  </View>
</View>
`;

exports[`Slider snapshots disabled = false, range = true 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 14,
        "height": 3.5,
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "left": 8.75,
            "width": 0,
          },
        }
      }
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 14,
          "height": 3.5,
          "left": 8.75,
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
  </View>
</View>
`;

exports[`Slider snapshots disabled = false, range = true, maxPointerValueInit = 80 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 14,
        "height": 3.5,
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "left": 8.75,
            "width": 0,
          },
        }
      }
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 14,
          "height": 3.5,
          "left": 8.75,
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="auto"
      style={
        {
          "backgroundColor": "#188700",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
  </View>
</View>
`;

exports[`Slider snapshots disabled = true, range = false 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 14,
        "height": 3.5,
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "left": 0,
            "width": 8.75,
          },
        }
      }
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 14,
          "height": 3.5,
          "left": 0,
          "position": "absolute",
          "width": 8.75,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="none"
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
  </View>
</View>
`;

exports[`Slider snapshots disabled = true, range = true 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 14,
        "height": 3.5,
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "left": 8.75,
            "width": 0,
          },
        }
      }
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 14,
          "height": 3.5,
          "left": 8.75,
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="none"
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="none"
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
  </View>
</View>
`;

exports[`Slider snapshots disabled = true, range = true, maxPointerValueInit = 80 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(0, 0, 0, 0.1000)",
        "borderRadius": 14,
        "height": 3.5,
        "justifyContent": "center",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "left": 8.75,
            "width": 0,
          },
        }
      }
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 14,
          "height": 3.5,
          "left": 8.75,
          "position": "absolute",
          "width": 0,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="none"
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
    <View
      collapsable={false}
      jestAnimatedStyle={
        {
          "value": {
            "transform": [
              {
                "translateX": 0,
              },
            ],
          },
        }
      }
      pointerEvents="none"
      style={
        {
          "backgroundColor": "grey",
          "borderRadius": 8.75,
          "height": 17.5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 17.5,
        }
      }
    />
  </View>
</View>
`;
